src/main/scala/scorex/transaction/NgHistoryReader.scala

```diff
@@ -0,0 +1,76 @@
+package scorex.transaction
+
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+import com.wavesplatform.state2._
+import scorex.block.Block.BlockId
+import scorex.block.{Block, MicroBlock}
+import scorex.transaction.History.BlockchainScore
+import scorex.utils.ScorexLogging
+
+class NgHistoryReader(ngState: () => Option[NgState], inner: History) extends History with NgHistory with DebugNgHistory with ScorexLogging with Instrumented {
+
+  override def synchronizationToken: ReentrantReadWriteLock = inner.synchronizationToken
+
+  override def height(): Int = read { implicit l =>
+    inner.height() + ngState().map(_ => 1).getOrElse(0)
+  }
+
+  override def blockBytes(height: Int): Option[Array[Byte]] = read { implicit l =>
+    inner.blockBytes(height).orElse(if (height == inner.height() + 1) ngState().map(_.bestLiquidBlock.bytes) else None)
+  }
+
+  override def scoreOf(blockId: BlockId): Option[BlockchainScore] = read { implicit l =>
+    inner.scoreOf(blockId)
+      .orElse(ngState() match {
+        case Some(ng) => Some(inner.score() + ng.base.blockScore)
+        case _ => None
+      })
+  }
+
+  override def heightOf(blockId: BlockId): Option[Int] = read { implicit l =>
+    lazy val innerHeight = inner.height()
+    inner.heightOf(blockId).orElse(ngState() match {
+      case Some(ng) if ng.contains(blockId) => Some(innerHeight + 1)
+      case _ => None
+    })
```

Comment:
I doesn't think that the version below is better, but keep in mind we can write something like:
```scala
ngState().filter(_.contains(blockId)).map(_ => innerHeight + 1)
```
;)